File: packages/rust/crates/oxbitnet-swift/Sources/OxBitNet/BitNet.swift
import Foundation
import COxBitNet

/// A loaded BitNet model. Thread-safe; blocking FFI calls are dispatched
/// to a background GCD queue to avoid starving the Swift cooperative pool.
public final class BitNet: @unchecked Sendable {
    private var handle: OpaquePointer?
    private let lock = NSLock()

    private init(handle: OpaquePointer) {
        self.handle = handle
    }

    deinit {
        dispose()
    }

    // MARK: - Loading

    /// Load a model asynchronously. The blocking C call runs on a background
    /// GCD queue so it won't block the Swift cooperative thread pool.
    public static func load(
        _ source: String,
        options: LoadOptions = LoadOptions()
    ) async throws -> BitNet {
        try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    let model = try loadSync(source, options: options)
                    continuation.resume(returning: model)
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Load a model synchronously (blocks the calling thread).
    public static func loadSync(
        _ source: String,
        options: LoadOptions = LoadOptions()
    ) throws -> BitNet {
        var cOpts = oxbitnet_default_load_options()
        var progressBox: Unmanaged<ProgressBox>?
        var cacheDirCString: UnsafeMutablePointer<CChar>?

        // Set up progress callback
        if let onProgress = options.onProgress {
            let box = ProgressBox(onProgress)
            let unmanaged = Unmanaged.passRetained(box)
            progressBox = unmanaged
            cOpts.on_progress = progressCallback
            cOpts.progress_userdata = unmanaged.toOpaque()
        }

        // Set up cache dir
        if let cacheDir = options.cacheDir {
            cacheDirCString = strdup(cacheDir)
            cOpts.cache_dir = UnsafePointer(cacheDirCString)
        }

        defer {
            progressBox?.release()
            free(cacheDirCString)
        }

        guard let handle = oxbitnet_load(source, &cOpts) else {
            let msg = errorMessage() ?? "unknown error"
            throw OxBitNetError.loadFailed(msg)
        }

        return BitNet(handle: handle)
    }

    // MARK: - Generate

    /// Generate tokens from a raw prompt. Returns an `AsyncThrowingStream`
    /// that yields tokens as they are produced.
    public func generate(
        _ prompt: String,
        options: GenerateOptions = GenerateOptions()
    ) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                do {
                    let handle = try acquireHandle()

                    var cOpts = OxBitNetGenerateOptions(
                        max_tokens: UInt(options.maxTokens),
                        temperature: options.temperature,
                        top_k: UInt(options.topK),
                        repeat_penalty: options.repeatPenalty,
                        repeat_last_n: UInt(options.repeatLastN)
                    )

                    let ctx = TokenStreamContext(continuation)
                    let unmanaged = Unmanaged.passRetained(ctx)
                    defer { unmanaged.release() }

                    let ret = oxbitnet_generate(
                        handle,
                        prompt,
                        &cOpts,
                        tokenCallback,
                        unmanaged.toOpaque()
                    )

                    if ret != 0 {
                        let msg = errorMessage() ?? "unknown error"
                        continuation.finish(throwing: OxBitNetError.generateFailed(msg))
                    } else {
                        continuation.finish()
                    }
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Chat

    /// Generate tokens from chat messages. Returns an `AsyncThrowingStream`
    /// that yields tokens as they are produced.
    public func chat(
        _ messages: [ChatMessage],
        options: GenerateOptions = GenerateOptions()
    ) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                do {
                    let handle = try acquireHandle()

                    var cOpts = OxBitNetGenerateOptions(
                        max_tokens: UInt(options.maxTokens),
                        temperature: options.temperature,
                        top_k: UInt(options.topK),
                        repeat_penalty: options.repeatPenalty,
                        repeat_last_n: UInt(options.repeatLastN)
                    )

                    let ctx = TokenStreamContext(continuation)
                    let unmanaged = Unmanaged.passRetained(ctx)
                    defer { unmanaged.release() }

                    // Build C message array — all strings must stay alive
                    // through the FFI call.
                    let cMessages: [OxBitNetChatMessage] = messages.map { msg in
                        OxBitNetChatMessage(
                            role: (msg.role as NSString).utf8String,
                            content: (msg.content as NSString).utf8String
                        )
                    }

                    let ret = cMessages.withUnsafeBufferPointer { buf in
                        oxbitnet_chat(
                            handle,
                            buf.baseAddress,
                            UInt(buf.count),
                            &cOpts,
                            tokenCallback,
                            unmanaged.toOpaque()
                        )
                    }

                    if ret != 0 {
                        let msg = errorMessage() ?? "unknown error"
                        continuation.finish(throwing: OxBitNetError.generateFailed(msg))
                    } else {
                        continuation.finish()
                    }
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Dispose

    /// Release all GPU resources. Safe to call multiple times.
    /// Also called automatically by `deinit`.
    public func dispose() {
        lock.lock()
        defer { lock.unlock() }
        if let h = handle {
            oxbitnet_free(h)
            handle = nil
        }
    }

    // MARK: - Private

    private func acquireHandle() throws -> OpaquePointer {
        lock.lock()
        defer { lock.unlock() }
        guard let h = handle else {
            throw OxBitNetError.disposed
        }
        return h
    }

    private static func errorMessage() -> String? {
        guard let ptr = oxbitnet_error_message() else { return nil }
        return String(cString: ptr)
    }
}

// Non-static helper for error messages after FFI calls on the same thread.
private func errorMessage() -> String? {
    guard let ptr = oxbitnet_error_message() else { return nil }
    return String(cString: ptr)
}

File: packages/rust/crates/oxbitnet-csharp/src/OxBitNet/BitNet.cs
using System;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace OxBitNet
{
    /// <summary>
    /// A loaded BitNet model. Thread-safe and disposable.
    /// </summary>
    public sealed class BitNet : IDisposable
    {
        private IntPtr _handle;
        private readonly object _lock = new object();
        private bool _disposed;

        static BitNet()
        {
            NativeLoader.EnsureRegistered();
        }

        private BitNet(IntPtr handle)
        {
            _handle = handle;
        }

        /// <summary>
        /// Load a model synchronously (blocks the calling thread).
        /// </summary>
        public static BitNet LoadSync(string source, LoadOptions options = null)
        {
            byte[] sourceBytes = Encoding.UTF8.GetBytes(source + '\0');
            var sourcePin = GCHandle.Alloc(sourceBytes, GCHandleType.Pinned);

            try
            {
                var cOpts = NativeMethods.oxbitnet_default_load_options();
                GCHandle progressPin = default;
                GCHandle cacheDirPin = default;
                OxBitNetProgressFn progressDelegate = null;

                try
                {
                    if (options?.OnProgress != null)
                    {
                        progressDelegate = (ref OxBitNetLoadProgress progress, IntPtr userdata) =>
                        {
                            var p = new LoadProgress(
                                (LoadPhase)progress.phase,
                                progress.loaded,
                                progress.total,
                                progress.fraction);
                            options.OnProgress(p);
                        };
                        progressPin = GCHandle.Alloc(progressDelegate);
                        cOpts.on_progress = Marshal.GetFunctionPointerForDelegate(progressDelegate);
                    }

                    if (options?.CacheDir != null)
                    {
                        byte[] cacheDirBytes = Encoding.UTF8.GetBytes(options.CacheDir + '\0');
                        cacheDirPin = GCHandle.Alloc(cacheDirBytes, GCHandleType.Pinned);
                        cOpts.cache_dir = cacheDirPin.AddrOfPinnedObject();
                    }

                    IntPtr handle = NativeMethods.oxbitnet_load(sourcePin.AddrOfPinnedObject(), ref cOpts);
                    if (handle == IntPtr.Zero)
                    {
                        throw new OxBitNetException(GetErrorMessage() ?? "Failed to load model");
                    }

                    return new BitNet(handle);
                }
                finally
                {
                    if (progressPin.IsAllocated) progressPin.Free();
                    if (cacheDirPin.IsAllocated) cacheDirPin.Free();
                }
            }
            finally
            {
                sourcePin.Free();
            }
        }

        /// <summary>
        /// Load a model asynchronously.
        /// </summary>
        public static Task<BitNet> Load(string source, LoadOptions options = null)
        {
            return Task.Run(() => LoadSync(source, options));
        }

        /// <summary>
        /// Generate text from a raw prompt string. Tokens are delivered via callback.
        /// </summary>
        public void Generate(string prompt, Action<string> onToken, GenerateOptions options = null)
        {
            IntPtr handle = AcquireHandle();
            var cOpts = MakeGenerateOptions(options);

            byte[] promptBytes = Encoding.UTF8.GetBytes(prompt + '\0');
            var promptPin = GCHandle.Alloc(promptBytes, GCHandleType.Pinned);

            OxBitNetTokenFn tokenDelegate = (IntPtr token, nuint len, IntPtr userdata) =>
            {
                string str = Marshal.PtrToStringUTF8(token, (int)len);
                onToken(str);
                return 0;
            };
            var tokenPin = GCHandle.Alloc(tokenDelegate);

            try
            {
                int ret = NativeMethods.oxbitnet_generate(
                    handle,
                    promptPin.AddrOfPinnedObject(),
                    ref cOpts,
                    Marshal.GetFunctionPointerForDelegate(tokenDelegate),
                    IntPtr.Zero);

                if (ret != 0)
                {
                    throw new OxBitNetException(GetErrorMessage() ?? "Generate failed");
                }
            }
            finally
            {
                promptPin.Free();
                tokenPin.Free();
            }
        }

        /// <summary>
        /// Generate text from a raw prompt string asynchronously.
        /// </summary>
        public Task GenerateAsync(string prompt, Action<string> onToken, GenerateOptions options = null)
        {
            return Task.Run(() => Generate(prompt, onToken, options));
        }

        /// <summary>
        /// Generate text from chat messages. Tokens are delivered via callback.
        /// </summary>
        public void Chat(ChatMessage[] messages, Action<string> onToken, GenerateOptions options = null)
        {
            IntPtr handle = AcquireHandle();
            var cOpts = MakeGenerateOptions(options);

            // Marshal chat messages
            var cMessages = new OxBitNetChatMessage[messages.Length];
            var pins = new GCHandle[messages.Length * 2];
            int pinIdx = 0;

            try
            {
                for (int i = 0; i < messages.Length; i++)
                {
                    byte[] roleBytes = Encoding.UTF8.GetBytes(messages[i].Role + '\0');
                    byte[] contentBytes = Encoding.UTF8.GetBytes(messages[i].Content + '\0');

                    var rolePin = GCHandle.Alloc(roleBytes, GCHandleType.Pinned);
                    var contentPin = GCHandle.Alloc(contentBytes, GCHandleType.Pinned);
                    pins[pinIdx++] = rolePin;
                    pins[pinIdx++] = contentPin;

                    cMessages[i].role = rolePin.AddrOfPinnedObject();
                    cMessages[i].content = contentPin.AddrOfPinnedObject();
                }

                var messagesPin = GCHandle.Alloc(cMessages, GCHandleType.Pinned);

                OxBitNetTokenFn tokenDelegate = (IntPtr token, nuint len, IntPtr userdata) =>
                {
                    string str = Marshal.PtrToStringUTF8(token, (int)len);
                    onToken(str);
                    return 0;
                };
                var tokenPin = GCHandle.Alloc(tokenDelegate);

                try
                {
                    int ret = NativeMethods.oxbitnet_chat(
                        handle,
                        messagesPin.AddrOfPinnedObject(),
                        (nuint)messages.Length,
                        ref cOpts,
                        Marshal.GetFunctionPointerForDelegate(tokenDelegate),
                        IntPtr.Zero);

                    if (ret != 0)
                    {
                        throw new OxBitNetException(GetErrorMessage() ?? "Chat failed");
                    }
                }
                finally
                {
                    messagesPin.Free();
                    tokenPin.Free();
                }
            }
            finally
            {
                for (int i = 0; i < pinIdx; i++)
                {
                    if (pins[i].IsAllocated) pins[i].Free();
                }
            }
        }

        /// <summary>
        /// Generate text from chat messages asynchronously.
        /// </summary>
        public Task ChatAsync(ChatMessage[] messages, Action<string> onToken, GenerateOptions options = null)
        {
            return Task.Run(() => Chat(messages, onToken, options));
        }

        /// <summary>
        /// Release all GPU resources. Safe to call multiple times.
        /// Also called automatically by Dispose().
        /// </summary>
        public void Dispose()
        {
            lock (_lock)
            {
                if (!_disposed && _handle != IntPtr.Zero)
                {
                    NativeMethods.oxbitnet_free(_handle);
                    _handle = IntPtr.Zero;
                    _disposed = true;
                }
            }
        }

        private IntPtr AcquireHandle()
        {
            lock (_lock)
            {
                if (_disposed || _handle == IntPtr.Zero)
                    throw new ObjectDisposedException(nameof(BitNet));
                return _handle;
            }
        }

        private static OxBitNetGenerateOptions MakeGenerateOptions(GenerateOptions options)
        {
            if (options == null)
                return NativeMethods.oxbitnet_default_generate_options();

            return new OxBitNetGenerateOptions
            {
                max_tokens = (nuint)options.MaxTokens,
                temperature = options.Temperature,
                top_k = (nuint)options.TopK,
                repeat_penalty = options.RepeatPenalty,
                repeat_last_n = (nuint)options.RepeatLastN,
            };
        }

        private static string GetErrorMessage()
        {
            IntPtr ptr = NativeMethods.oxbitnet_error_message();
            if (ptr == IntPtr.Zero) return null;
            return Marshal.PtrToStringUTF8(ptr);
        }
    }
}
